sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/BaseAuthorizationTokenProvider.java

```diff
@@ -240,24 +230,43 @@ public String generateKeyAuthorizationSignatureNew(String verb, String resourceI
         // If any of the value is optional, it should still have the placeholder value
         // of ""
         // OperationType -> ResourceType -> ResourceId/OwnerId -> XDate -> Date
-        String verbInput = verb;
-        String resourceIdInput = resourceIdValue;
-        String resourceTypeInput = resourceType;
-
-        String authResourceId = getAuthorizationResourceIdOrFullName(resourceTypeInput, resourceIdInput);
-        String payLoad = generateMessagePayload(verbInput, authResourceId, resourceTypeInput, headers);
-        Mac mac = null;
-        try {
-            mac = (Mac) this.macInstance.clone();
-        } catch (CloneNotSupportedException e) {
-            throw new IllegalStateException(e);
-        }
+
+        String authResourceId = getAuthorizationResourceIdOrFullName(resourceType, resourceIdValue);
+        String payLoad = generateMessagePayload(verb, authResourceId, resourceType, headers);
+        Mac mac = this.getMacInstance();
         byte[] digest = mac.doFinal(payLoad.getBytes());
         String authorizationToken = Utils.encodeBase64String(digest);
         String authtoken = AUTH_PREFIX + authorizationToken;
         return HttpUtils.urlEncode(authtoken);
     }
 
+    private Mac getMacInstance() {
+        int masterKeyLatestHashCode = this.cosmosKeyCredential.getMasterKey().hashCode();
```

Comment:
you are evaluating masterkey#hashCode on every single call.
And seems your intention is to only use it for comparing the new hashcode with the old hashcode.

String comparison directly should be faster that hashcode computation. As for hashcode computation you are iteratring over all chars of the string and doing some arithmatic but in string comparison you just comparing the chars.

Reply:
Good point, but the reason for hashCode comparison was because of the fact that we don't want to store masterKey in BaseAuthorizationTokenProvider, because we always want to use cosmosKeyCredential reference to get the master key. 
But as you mentioned, string comparison is faster, I can store the master key and won't use it for any other thing than comparison. Makes sense ?

Reply:
do no-op on this. Thanks for the explanation.